### PR TITLE
feat(multi-part)

### LINKED ISSUE #44: Multi-part Body Upload
Enable multi-part body upload for:
- [ ] Automatic Speech Recognition.
- [ ] Image Classification
- [ ] Zero-Shot Image Classification
- [ ] Object Detection

### PR DESCRIPTION
**Pull request by ovuruska**

### CODE REVIEW

File: src/clients/deep-infra.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from "axios";
 import { USER_AGENT } from "@/lib/constants/client";
 import { ClientConfig, IClientConfig } from "@/lib/types/common/client-config";
+import FormData from "form-data";
```
Comment (ichernev) [inline src/clients/deep-infra.ts]: Why is this needed here?

File: src/clients/deep-infra.ts [hunk 2]
```diff
@@ -30,15 +31,17 @@ export class DeepInfraClient {
     config?: AxiosRequestConfig,
   ): Promise<AxiosResponse<T>> {
     const headers = {
+      "User-Agent": USER_AGENT,
       "Content-Type": "application/json",
+      ...config?.headers,
```
Comment (ichernev) [inline src/clients/deep-infra.ts:34]: So the reason the headers are expanded here is so that they can override the UA and CT but not the Authorization? Is there a reason for this decision? I mean why not?

File: src/lib/models/base/image-base-model.ts [hunk 3]
```diff
@@ -24,4 +25,16 @@ export class ImageBaseModel<
     });
     return response.data;
   }
+
+  async generateWithFormData(body: RequestType): Promise<ResponseType> {
```
Comment (ichernev) [inline src/lib/models/base/image-base-model.ts]: The way I was imagining it, we should pick the right body (i.e json or multi-part) in code (maybe hardcoded for some request types, maybe dynamic based on provided arguments (i.e if the user provides base64, send as json, otherwise send as multipart).

But not the user having to decide how to send it (because in the end of the day this is an implementation detail).

I still don't see what possible options are supported (other than fs.createReadStream being able to work on them).

Comment (ichernev) [reply src/lib/models/base/image-base-model.ts]: Also, for some requests, there are other arguments (in addition to the image), so those have to also be sent as another `part` in the multipart body (and encoded properly, there it gets a bit dicey). E2E testing should help here :)

**Comment by ovuruska** [comment]
Unit test development is still in progress.

File: src/lib/models/base/image-base-model.ts [hunk 4]
```diff
@@ -16,11 +17,17 @@ export class ImageBaseModel<
...
-    const response = await this.client.post<ResponseType>({
-      ...body,
-      image: base64Content,
+    const { image , ...rest} = body;
+    const formData = new FormData();
+    const readStream = await ReadStreamUtils.getReadStream(image);
+    formData.append("image", readStream);
+    Object.entries(rest).forEach(([key, value]) => {
+      formData.append(key, JSON.stringify(value));
+    });
```
Comment (ichernev) [inline src/lib/models/base/image-base-model.ts]: You can probably wrap this in a utility method, that receives "blob-like" key names, and treats the rest via JSON.stringify. One reason is to reduce duplication, another is that I'm worried the JSON.stringify is not the right option, but if it's in one place it would be easier to manage.

I assume you have tested this on all affected categories with some non-blob inputs?

Comment (dinsizimansiz) [reply src/lib/models/base/image-base-model.ts:26]: Gotcha. I'm going to wrap form data serialization under one utility class.

Comment (dinsizimansiz) [reply src/lib/models/base/image-base-model.ts:26]: I tested image input from URL, from file and from buffer. Blob should also be included.

File: test/base/automatic-speech-recognition.spec.ts [hunk 5]
```diff
@@ -48,4 +42,16 @@ describe("AutomaticSpeechRecognition", () => {
       expect.any(Object),
     );
   });
+
+  it("should send a request with base64 audio", async () => {
```
Comment (ichernev) [inline test/base/automatic-speech-recognition.spec.ts]: I guess this text is not accurate any more :) I think we always send FormData now (for certain types of models).

File: test/base/zero-shot-image-classification.spec.ts [hunk 6]
```diff
@@ -47,4 +40,17 @@ describe("ZeroShotImageClassification", () => {
...
     );
   });
+
+  it("should send a request with base64 image", async () => {
+    const response = await model.generate({
+      image: fakeFileBuffer as Buffer,
+      candidate_labels: ["dog", "cat"],
+    });
+    expect(response).toBeDefined();
+    expect(postMock).toHaveBeenCalledWith(
```
Comment (ichernev) [inline test/base/zero-shot-image-classification.spec.ts]: I guess you can do a better job at asserting here something more meaningful. Also the test name is bogus.
